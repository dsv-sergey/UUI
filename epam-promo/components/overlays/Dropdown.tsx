import * as React from 'react';
import { Dropdown as UuiDropdown, DropdownProps } from '@epam/uui-components';

export class Dropdown extends React.Component<DropdownProps, any> {
    render() {
        return (
            <UuiDropdown  { ...this.props } />
        );
    }
}
