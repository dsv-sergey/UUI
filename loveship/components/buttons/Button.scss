@use '../../assets/styles/scss/index' as *;

.root {
    @include button-reset();
    font-family: $font-sans-semibold;
    font-weight: normal;

    &:global(.uui-button-box) {
        border-style: solid;
        justify-content: center;
        border-radius: $control-border-radius;
    }

    :global(.uui-caption) {
        @include overflow-ellipsis();
        min-width: 0;
        text-align: center;
        flex: 1 1 auto;
    }

    &:global(.-clickable):focus {
        @include control-focus-effect();
    }
}

// Sizes
.size-18 { @include button-layout(18px, 1px); }
.size-24 { @include button-layout(24px, 1px); }
.size-30 { @include button-layout(30px, 1px); }
.size-36 { @include button-layout(36px, 1px); }
.size-42 { @include button-layout(42px, 1px); }
.size-48 { @include button-layout(48px, 1px); }

// Border style
.style-round {
    &.size-48 { border-radius: 24px; }
    &.size-42 { border-radius: 21px; }
    &.size-36 { border-radius: 18px; }
    &.size-30 { border-radius: 15px; }
    &.size-24 { border-radius: 12px; }
    &.size-18 { border-radius: 10px; }
}

.fill-solid {
    @include button-colors(var(--uui-color-700), var(--uui-contrast-color-700), var(--uui-color-700));

    &:global(.-clickable) {
        &:hover {
            @include button-colors(var(--uui-color-800), var(--uui-contrast-color-700), var(--uui-color-800));
        }

        &:active {
            @include button-colors(var(--uui-color-900), var(--uui-contrast-color-700), var(--uui-color-900));
        }
    }

    &:global(.uui-disabled) {
        @include button-colors($night300, $night500, $night300);
        cursor: default;

        &:hover, &:active, &:focus {
            @include button-colors($night300, $night500, $night300);
            box-shadow: none;
        }

        &::placeholder {
            color: $night300;
        }
    }
}

.fill-white {
    @include button-colors(white, var(--uui-color-700), var(--uui-color-700));

    &:global(.-clickable) {
        &:hover {
            @include button-colors(var(--uui-color-500), var(--uui-color-800), var(--uui-color-800));
        }

        &:active {
            @include button-colors(var(--uui-color-600), var(--uui-color-900), var(--uui-color-900));
        }
    }

    &:global(.uui-disabled) {
        @include button-colors(white, $night500, $night300);
        cursor: default;

        &:hover, &:active, &:focus {
            @include button-colors(white, $night500, $night300);
            box-shadow: none;
        }

        &::placeholder {
            color: $night300;
        }
    }
}

.fill-none {
    @include button-colors(transparent, var(--uui-color-700), var(--uui-color-700));

    &:global(.-clickable) {
        &:hover {
            @include button-colors(var(--uui-color-500), var(--uui-color-800), var(--uui-color-800));
        }

        &:active {
            @include button-colors(var(--uui-color-600), var(--uui-color-900), var(--uui-color-900));
        }
    }

    &:global(.uui-disabled) {
        @include button-colors(transparent, $night500, $night300);
        cursor: default;

        &:hover, &:active, &:focus {
            @include button-colors(transparent, $night500, $night300);
            box-shadow: none;
        }

        &::placeholder {
            color: $night300;
        }
    }
}

.fill-light {
    @include button-colors(transparent, var(--uui-color-700), transparent);

    &:global(.-clickable) {
        &:hover {
            @include button-colors(var(--uui-color-500), var(--uui-color-800), var(--uui-color-500));
        }

        &:active {
            @include button-colors(var(--uui-color-600), var(--uui-color-900), var(--uui-color-600));
        }
    }

    &:global(.uui-disabled) {
        @include button-colors(transparent, $night500, transparent);
        cursor: default;

        &:hover, &:active, &:focus {
            @include button-colors(transparent, $night500, transparent);
            box-shadow: none;
        }

        &::placeholder {
            color: $night300;
        }
    }
}
