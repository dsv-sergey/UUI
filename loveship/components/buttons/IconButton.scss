@use '../../assets/styles/scss/index'  as *;

.root {
    @include button-reset();
    overflow: visible;
    cursor: pointer;
    fill: var(--uui-color-700);

    &:global(.-clickable) {
        &:hover {
            fill: var(--uui-color-800);
        }

        &:active {
            fill: var(--uui-color-900);
        }
    }

    &:global(.uui-disabled) {
        fill: $night500;
        cursor: default;

        &:hover {
            fill: $night500;
        }
    }

    &:global(.uui-invalid) {
        fill: $fire;
    }
}
