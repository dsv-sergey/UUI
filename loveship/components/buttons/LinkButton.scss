@use '../../assets/styles/scss/index'  as *;

.root {
    @include button-reset();

    // Fonts
    @include epam-uui-fonts();

    // Sizes
    &.size-18 {
        @include link-button-layout(18px);
        min-width: auto;
    }

    &.size-24 {
        @include link-button-layout(24px);
        min-width: auto;
    }

    &.size-30 {
        @include link-button-layout(30px);
        min-width: auto;
    }

    &.size-36 {
        @include link-button-layout(36px);
        min-width: auto;
    }

    &.size-42 {
        @include link-button-layout(42px);
        min-width: auto;
    }

    &.size-48 {
        @include link-button-layout(48px);
        min-width: auto;
    }

    &:global(.uui-button-box) {
        @include button-colors(transparent, var(--uui-color-700), transparent);

        &:global(.-clickable) {
            &:hover {
                @include button-colors(transparent, var(--uui-color-800), transparent);
            }

            &:active {
                @include button-colors(transparent, var(--uui-color-900), transparent);
            }

            &:focus {
                box-shadow: none;
            }
        }

        &:global(.uui-disabled) {
            @include disabled-input();
            background-color: transparent;

            &:hover {
                background-color: transparent;
            }
        }
    }
}
