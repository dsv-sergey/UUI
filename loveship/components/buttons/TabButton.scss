@use '../../assets/styles/scss/index'  as *;

$text-size-caption: 36px;
$text-size-count: 24px;

@mixin tab-button-color($bg, $fg, $brd) {
    color: $fg;
    fill: $fg;

    &:hover {
        fill: $bg;
        color: $bg;
    }

    &:global(.uui-active) {
        fill: $bg;
        color: $bg;
        border-color: $brd;
    }
}

.root {
    @include button-reset();
    @include link-button-horizontal-layout();
    min-width: 36px;
    padding-left: 12px;
    padding-right: 12px;
    position: relative;
    bottom: -1px;
    display: flex;
    align-items: center;
    font-family: $font-sans-semibold;
    font-weight: normal;
    border-bottom: 2px solid transparent;

    // &:focus {
    //     .control-focus-effect();
    // }

    &.theme-light {
        @include tab-button-color($sky, $night700, $sky);

        &:global(.uui-disabled) {
            @include tab-button-color($night500, $night500, $night500);
        }
    }

    &.theme-dark {
        @include tab-button-color($sky, white, $sky);

        &:global(.uui-disabled) {
            @include tab-button-color($night600, $night600, $night600);
        }
    }

    &.uui-notification {
        &::after {
            content: '';
            margin-bottom: $text-size-caption / 2;
            height: 6px;
            width: 6px;
            background: $fire;
            box-shadow: 0 1px 8px 0 $fire;
            border-radius: 50%;
        }
    }

    :global(.uui-count) {
        @include text-size($text-size-count, 0);
        letter-spacing: 0;
        color: $night600;
        background-color: $night300;
        padding: 0 6px;
        margin-left: 6px;
        margin-right: 3px;
        border-radius: 9px;
    }

    &:global(.uui-active) {
        border-bottom: 2px solid;
    }

    :global(.uui-caption) {
        @include text-size($text-size-caption, 0);
    }

    // Sizes
    &.size-36 {
        height: 36px;
    }

    &.size-48 {
        height: 48px;
    }

    &.size-60 {
        height: 60px;
    }
}
