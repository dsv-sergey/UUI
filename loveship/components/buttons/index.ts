export * from './Button';
export * from './LinkButton';
export * from './IconButton';
export * from './TabButton';
export * from './helper';