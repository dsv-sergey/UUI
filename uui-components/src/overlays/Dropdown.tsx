import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Manager, Reference, Popper, ReferenceChildrenProps } from 'react-popper';
import { Portal } from './Portal';
import { isClickableChildClicked, IEditable, LayoutLayer, IDropdownToggler, UuiContexts, closest } from '@epam/uui';
import * as PopperJS from 'popper.js';
import * as PropTypes from 'prop-types';
import { Ref } from "react";
import { PopperTargetWrapper } from "./PopperTargetWrapper";

export interface DropdownState {
    opened: boolean;
    bodyBoundingRect: { y: number | null; x: number | null, width: number | null, height: number | null };
    closeDropdownTimerId: any;
}

export interface DropdownBodyProps {
    onClose(): void;
    togglerWidth: number;
    togglerHeight: number;
    scheduleUpdate: () => void;
}

export type DropdownPlacement = PopperJS.Placement;

export interface DropdownProps extends Partial<IEditable<boolean>> {
    renderTarget: (props: IDropdownToggler & {ref?: Ref<any>}) => React.ReactNode;
    renderBody: (props: DropdownBodyProps & any) => React.ReactNode;
    onClose?: () => any;
    isNotUnfoldable?: boolean;
    stopCloseSelectors?: string[];
    zIndex?: number;
    placement?: DropdownPlacement;
    modifiers?: PopperJS.Modifiers;
    /** Should we close dropdown on click on the Toggler, if it's already open? Default is true. */

    openOnClick?: boolean; // default: true
    openOnHover?: boolean; // default: false
    closeOnTargetClick?: boolean; // default: true
    closeOnClickOutside?: boolean; // default: true
    closeOnMouseLeave?: 'toggler' | 'boundary' | false;

    portalTarget?: HTMLElement;
    boundaryElement?: PopperJS.Boundary | Element;
}

export class Dropdown extends React.Component<DropdownProps, DropdownState> {
    private targetNode: HTMLElement | null = null;
    private bodyNode: HTMLElement | null = null;
    private lastOpenedMs: number;
    private togglerWidth: number;
    private togglerHeight: number;

    state: DropdownState = {
        opened: this.props.value || false,
        bodyBoundingRect: { y: null, x: null, height: null, width: null },
        closeDropdownTimerId: null,
    };

    static contextTypes = {
        uuiLayout: PropTypes.object,
    };

    private readonly layer: LayoutLayer | null = null;

    constructor(props: DropdownProps, public context: UuiContexts) {
        super(props, context);
        this.layer = this.context.uuiLayout && this.context.uuiLayout.getLayer();
    }

    public componentDidMount() {
        window.addEventListener('dragstart', this.clickOutsideHandler);
        window.addEventListener('scroll', this.clickOutsideHandler, true);

        if (this.props.openOnHover && !this.props.openOnClick) {
            this.targetNode?.addEventListener('mouseenter', this.handleMouseEnter);
        }
        if (this.props.closeOnMouseLeave === 'toggler') {
            this.targetNode?.addEventListener('mouseleave', this.handleMouseLeave);
        }
        if (this.props.closeOnClickOutside !== false) {
            window.addEventListener('click', this.clickOutsideHandler, true);
        }

    }

    public componentWillUnmount() {
        window.removeEventListener('dragstart', this.clickOutsideHandler);
        window.removeEventListener('scroll', this.clickOutsideHandler, true);
        window.removeEventListener('click', this.clickOutsideHandler, true);

        this.context.uuiLayout?.releaseLayer(this.layer);
    }

    handleOpenedChange = (opened: boolean) => {
        if (opened && this.props.closeOnMouseLeave === 'boundary') {
            window.addEventListener('mousemove', this.handleMouseMove);
        } else if (!opened && this.props.closeOnMouseLeave === 'boundary') {
            window.removeEventListener('mousemove', this.handleMouseMove);
        }

        this.props.onValueChange ? this.props.onValueChange(opened) : this.setState({ opened: opened });
        if (opened) {
            this.lastOpenedMs = (new Date()).getTime();
        }
    }

    isOpened = () => {
        return this.props.value !== undefined ? this.props.value : this.state.opened;
    }

    private handleTargetClick = (e: Event) => {
        if (!this.props.isNotUnfoldable && !(e && isClickableChildClicked(e as any))) {
            const currentValue = this.props.value !== undefined ? this.props.value : this.state.opened;
            const newValue = (this.props.closeOnTargetClick === false) ? true : !currentValue;
            if (currentValue !== newValue) {
                this.handleOpenedChange(newValue);
            }
        }
    }

    private handleMouseEnter = (e: Event) => {
        this.handleOpenedChange(true);
    }

    private handleMouseLeave = (e: MouseEvent) => {
        this.handleOpenedChange(false);
    }

    isClientInArea(e: MouseEvent) {
        const areaPadding = 30;
        const { y, x, height, width } = this.state.bodyBoundingRect;

        if (y && x && width && height) {
            return x - areaPadding <= e.clientX && e.clientX <= x + areaPadding + width && y - areaPadding <= e.clientY && e.clientY <= y + height + areaPadding;
        }
    }

    setCloseDropdownTimer() {
        this.setState({
            closeDropdownTimerId: setTimeout(() => { this.handleOpenedChange(false); this.clearCloseDropdownTimer(); }, 1500),
        });
    }

    clearCloseDropdownTimer() {
        if (this.state.closeDropdownTimerId) {
            clearTimeout(this.state.closeDropdownTimerId);
            this.setState({closeDropdownTimerId: null});
        }
    }

    private handleMouseMove = (e: MouseEvent) => {
        if (this.isInteractedOutside(e) && !this.isClientInArea(e)) {
            this.handleOpenedChange(false);
            this.clearCloseDropdownTimer();
        } else if (this.isInteractedOutside(e) && !this.state.closeDropdownTimerId) {
            this.setCloseDropdownTimer();
        } else if (!this.isInteractedOutside(e) && this.state.closeDropdownTimerId) {
            this.clearCloseDropdownTimer();
        }
    }

    private onClose = () => {
        this.props.onClose ? this.props.onClose() : this.setState({ opened: false });
    }

    private renderTarget(targetProps: ReferenceChildrenProps) {
        const innerRef = (node: Element) => {
            this.targetNode = ReactDOM.findDOMNode(node) as HTMLElement;
            targetProps.ref(this.targetNode);
        };

        return (
            <PopperTargetWrapper innerRef={ innerRef }>
            {
                this.props.renderTarget({
                    onClick: (this.props.openOnClick || (!this.props.openOnClick && !this.props.openOnHover)) ? this.handleTargetClick : undefined,
                    isOpen: this.isOpened(),
                    isDropdown: true,
                })
            }
            </PopperTargetWrapper>
        );
    }

    private isInteractedOutside = (e: Event) => {
        if (!this.isOpened()) {
            return false;
        }

        // Opening the dropdown often cause Search control inside to auto-focus
        // That's (for not investigated reason) causes onScroll event on window,
        //   which in turn cause the dropdown to close immediately after opening
        // This hack prevents this
        // This hack also fixes scroll event on mobile when focus on input
        if (e.type === 'scroll' && ((new Date()).getTime() - this.lastOpenedMs) < 1000) {
            return false;
        }

        const target = (e.target as HTMLElement);
        const stopNodes = [this.bodyNode, this.targetNode, ...(this.props.stopCloseSelectors || [])];

        if (stopNodes.some((node: any) => node && closest(target, node))) {
            return false;
        }

        if (closest(target, '.uui-popper') && +closest(target, '.uui-popper').style.zIndex > (this.bodyNode !== null ? +this.bodyNode.style.zIndex : 0)) {
            return false;
        }

        return true;
    }

    private clickOutsideHandler = (e: Event) => {
        if (this.isInteractedOutside(e)) {
            this.handleOpenedChange(false);
        }
    }

    private updateTogglerSize() {
        if (this.targetNode) {
            const rect = this.targetNode.getBoundingClientRect();
            this.togglerWidth = rect.width;
            this.togglerHeight = rect.height;
        }
    }

    public render() {
        const shouldShowBody = this.isOpened() && !this.props.isNotUnfoldable;

        if (shouldShowBody) {
            this.updateTogglerSize();
        }

        return (
            <Manager>
                <Reference>
                    { (targetProps: ReferenceChildrenProps) => this.renderTarget(targetProps) }
                </Reference>
                { shouldShowBody && <Portal target={ this.props.portalTarget }>
                    <Popper
                        placement={ this.props.placement || 'bottom-start' }
                        positionFixed={ true }
                        modifiers={ {
                            ...this.props.modifiers,
                            preventOverflow: {
                                boundariesElement: this.props.boundaryElement || 'viewport',
                                ...((this.props.modifiers || {}).preventOverflow || {}),
                            },
                        } }
                    >
                        { ({ ref, style, placement, scheduleUpdate }) =>
                            <div
                                className='uui-popper'
                                ref={ node => {
                                    ref(node);
                                    this.bodyNode = node;
                                    if (this.bodyNode && this.props.closeOnMouseLeave === 'boundary') {
                                        const { x, y, height, width } = this.bodyNode.getBoundingClientRect() as DOMRect;
                                        if (x && y && !this.state.bodyBoundingRect.y && !this.state.bodyBoundingRect.x) {
                                            this.setState({ bodyBoundingRect : { y, height, width, x } });
                                        }
                                    }
                                } }
                                style={ { ...style, zIndex: this.props.zIndex != null ? this.props.zIndex : this.layer?.zIndex } }
                                data-placement={ placement }
                            >
                                { this.props.renderBody({
                                    onClose: this.onClose,
                                    togglerWidth: this.togglerWidth,
                                    togglerHeight: this.togglerHeight,
                                    scheduleUpdate,
                                }) }
                            </div>
                        }
                    </Popper>
                </Portal> }
            </Manager>
        );
    }

}
