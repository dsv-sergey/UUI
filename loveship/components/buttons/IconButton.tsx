import * as React from 'react';
import { ButtonBaseProps, ButtonBase } from '@epam/uui-components';
import { IconContainer } from '@epam/uui-components';
import * as types from '../types';
import * as css from './IconButton.scss';
import * as styles from '../../assets/styles/scss/loveship-color-vars.scss';

export interface IconButtonProps extends ButtonBaseProps, types.ColorMod { }

export class IconButton extends ButtonBase<IconButtonProps> {
    constructor(props: IconButtonProps) {
        super(props);
    }

    getClassName() {
        return [
            css.root,
            styles['color-' + (this.props.color || 'night600')],
        ];
    }

    getChildren() {
        return [
            <IconContainer key={ 'icon' } icon={ this.props.icon } />,
        ];
    }
}