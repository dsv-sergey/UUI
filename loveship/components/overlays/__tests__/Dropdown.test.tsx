import React from 'react';
import renderer from 'react-test-renderer';
import { Dropdown } from '../Dropdown';
import { Button } from '../../buttons';

jest.mock('react-dom', () => ({
    findDOMNode: jest.fn(),
}));

describe('Dropdown', () => {
    it('should be rendered correctly', () => {
        const tree = renderer
            .create(<Dropdown
                renderTarget={ props => <Button caption='Test' { ...props } /> }
                renderBody={  jest.fn() }
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly with extra props', () => {
        const tree = renderer
            .create(<Dropdown
                renderTarget={ props => <Button caption='test' { ...props } /> }
                renderBody={ jest.fn() }
                onClose={ jest.fn() }
                stopCloseSelectors={ ['test-selector'] }
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});


