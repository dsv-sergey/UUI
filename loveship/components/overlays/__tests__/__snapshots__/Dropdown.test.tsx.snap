// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should be rendered correctly 1`] = `
<div
  className="container uui-button-box uui-enabled -clickable root size-36 color-sky fill-solid style-square"
  href={null}
  onClick={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <div
    className="uui-caption line-height-24 font-size-14 v-padding-5"
  >
    Test
  </div>
  <div
    className="container uui-icon uui-enabled"
  >
    <svg
      className=""
      height={24}
      viewBox="0 0 12 24"
      width={12}
    >
      <use
        xlinkHref="#folding-arrow-18.svg"
      />
    </svg>
  </div>
</div>
`;

exports[`Dropdown should be rendered correctly with extra props 1`] = `
<div
  className="container uui-button-box uui-enabled -clickable root size-36 color-sky fill-solid style-square"
  href={null}
  onClick={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <div
    className="uui-caption line-height-24 font-size-14 v-padding-5"
  >
    test
  </div>
  <div
    className="container uui-icon uui-enabled"
  >
    <svg
      className=""
      height={24}
      viewBox="0 0 12 24"
      width={12}
    >
      <use
        xlinkHref="#folding-arrow-18.svg"
      />
    </svg>
  </div>
</div>
`;
